### PR TITLE
ensemble model parallelism

### LINKED ISSUE #1341: [TASK] Ensembe model improve parallelism strategy in case of no codes involved
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
In case of ensemble modeling not involving codes, the ensemble model should use the standard parallelization strategy (no Client mode) since, if large datasets need to be transfered from a model to the other, the client approach can deteriorate the parallelism

**Describe the solution you'd like**
If no Codes are involeved in  an ensemble model, the parallelism should follow a standard "MPI-like" approach

**Describe alternatives you've considered**
 N.A

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes  #1341

##### What are the significant changes in functionality due to this change request?
Changed the parallelism strategy in case of ensemble model.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).  (No changes to input syntax)
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. (Yes)
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. (yes)
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. (no significant functionality added)
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.  (needs second)
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. (done)
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? (no change)
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example. (no change)

### CODE REVIEW

**Comment by alfoa** [comment]
@joshua-cogliati-inl ready to review

File: framework/Models/EnsembleModel.py
Comment (joshua-cogliati-inl) [inline framework/Models/EnsembleModel.py:727]: I think this is a poor use of the jobHandler, to run only one input, since we could just run it directly without taking up a second jobhandler slot.  This probably at least deserves a FIXME comment.  (Alternatively, maybe we could remove the else part of  ```self.parallelStrategy == 1``` ?)

File: framework/Models/EnsembleModel.py
Comment (joshua-cogliati-inl) [inline framework/Models/EnsembleModel.py:723]: I think (unless I am misreading the code), nextModel and the ```while not nextModel``` loop could be removed and this would simplify the code.

File: framework/SupervisedLearning/ARMA.py
Comment (joshua-cogliati-inl) [inline framework/SupervisedLearning/ARMA.py:731]: Can this comment be removed? (Or should it be noted that it is for an older statsmodel?)

File: framework/utils/randomUtils.py
Comment (joshua-cogliati-inl) [inline framework/utils/randomUtils.py:157]: dim is nolonger a parameter.

File: framework/utils/randomUtils.py
Comment (joshua-cogliati-inl) [inline framework/utils/randomUtils.py:185]: optional, optional?

File: framework/utils/randomUtils.py
Comment (joshua-cogliati-inl) [inline framework/utils/randomUtils.py:194]: Is round correct here, or would floor or ceil be better? I'm worried that it is biasing it away from some values.

File: tests/framework/CodeInterfaceTests/Instant/IAEA2D/sample/1/alias.xml
Comment (joshua-cogliati-inl) [inline tests/framework/CodeInterfaceTests/Instant/IAEA2D/sample/1/alias.xml:1]: Is IAEA2D/sample/1 a generated directory (as in should it be in the repository)?

**Review by joshua-cogliati-inl — CHANGES_REQUESTED** [review]
There are some things that need to be fixed. I would prefer if this was split into two pull requests, but that is not necessary.  I also had some questions.  

There are definitely improvements in this pull request :)

File: framework/SupervisedLearning/ARMA.py
Comment (PaulTalbot-INL) [reply framework/SupervisedLearning/ARMA.py:731]: I would appreciate leaving this here at least until the new changes can be tested in production a bit more, for reference.

File: framework/Models/EnsembleModel.py
Comment (alfoa) [reply framework/Models/EnsembleModel.py:298]: ops.... typo hahaha

File: framework/Models/EnsembleModel.py
Comment (alfoa) [reply framework/Models/EnsembleModel.py:723]: you are right...it was a left over of an old implementation...fixed

File: framework/Models/EnsembleModel.py
Comment (alfoa) [reply framework/Models/EnsembleModel.py:727]: Agree. I modified it a bit to submit directly in the internal queue and just wait outside.... I really want to eliminate this internal job handler...I think we need to have a discussion about how to handle "MPI" within a thread (ray thread)

File: framework/utils/randomUtils.py
Comment (alfoa) [reply framework/utils/randomUtils.py:194]: Hi Josh. I think it should be fine here since the random returns a float between 0 1

File: tests/framework/CodeInterfaceTests/Instant/IAEA2D/sample/1/alias.xml
Comment (alfoa) [reply tests/framework/CodeInterfaceTests/Instant/IAEA2D/sample/1/alias.xml]: I fixxed and removed all the un-useful files there (sorry for this...good catch)

**Comment by alfoa** [comment]
@joshua-cogliati-inl Josh I addressed your comments

**Comment by PaulTalbot-INL** [comment]
Any hope for a end-of-week miracle merge on this?

**Review by joshua-cogliati-inl — APPROVED** [review]

**Comment by alfoa** [comment]
@PaulTalbot-INL looks like Josh is on it...so you called the miracle ehhehe

**Comment by joshua-cogliati-inl** [comment]
I approve, but this changes a requirements test so needs a second approval from the CCB chair or designee.

**Comment by joshua-cogliati-inl** [comment]
@PaulTalbot-INL @alfoa This changes a requirements test, so it needs a second approver.

**Comment by alfoa** [comment]
As chair of COB I approve the requirement test modifications.

**Comment by alfoa** [comment]
@PaulTalbot-INL  can you also approve the PR?

File: framework/Models/EnsembleModel.py
Comment (PaulTalbot-INL) [inline framework/Models/EnsembleModel.py:502]: Is this just `kwargsToKeep == copy.deepcopy(kwargs)`? No change needed, I see how this could be useful in the future if we have keys we don't want to keep.

**Review by PaulTalbot-INL — APPROVED** [review]
I'm okay with all these changes, as well.

File: framework/Models/EnsembleModel.py
Comment (alfoa) [reply framework/Models/EnsembleModel.py:502]: ops...yes.. i will address this in next PR

**Pull request merged**
